### DISCUSSION TITLE
Split revisions by op

### DISCUSSION THREAD

**Discussion by Andrew-Chen-Wang** [sqlalchemy/alembic#1398] [Usage Questions] [answered]
Hi I'm looking to split revisions by op due to scaling issues. For instance:

```
def upgrade():
    op.add_column()
    op.add_column()

def downgrade():
    op.drop_column()
    op.drop_column()
```

Should be split into two files. Guidance would be great! Currently, I just use a pytest to check migration files using ast. I feel like this would be harder as a post-migration step trying to match downgrade operations with upgrade operations.

Actually, question: for every upgrade operation, is there a parallel downgrade function? If the number of operations per function is the same, then I can go about it like that.

**Comment by CaselIT** [comment]
Hi,

That's a strange request, I'm not sure if there is something that can do that automatically. 
I suggest you take a look at the api exposed here https://alembic.sqlalchemy.org/en/latest/api/autogenerate.html#customizing-revision-generation

**Reply by Andrew-Chen-Wang** [reply]
Thanks, the API doc is helpful. Off the top of your head, for every upgrade operation, is there a parallel downgrade function? i.e. # of upgrade ops = # of downgrade ops?

**Reply by CaselIT** [reply, marked as answer]
I would say that in most cases you could assume that.
The only cases that I can think of would be an add table with complex fk/index/etc that can't be added inline requiring multiple operations to create and only a drop table to undo. But I'm not sure autogenerate will generate that or create a single create_table even in this case.

this may not be the only instance though.

I guess add an assert that len(upgrade_ops) == len(downgrade_ops) and see how far you get

**Reply by Andrew-Chen-Wang** [reply]
Thanks for the guidance. I'm pretty satisfied with the results. I would publish the script I've written, but, for future readers, it's a lot of ast parsing, unparsing, the NodeTransforming that I simply generated with ChatGPT ([link](https://chat.openai.com/share/dab48ee8-76d6-4b01-a3f7-3d25d246a30b)).
